File: lib/models/course.dart
class Course {
  final String id;
  final String code;
  final String name;
  final String instructor;
  final String schedule;
  final String location;
  final int credits;
  final String description;
  final List<String> enrolledStudentIds;
  final List<Channel> channels;

  const Course({
    required this.id,
    required this.code,
    required this.name,
    required this.instructor,
    required this.schedule,
    required this.location,
    required this.credits,
    required this.description,
    this.enrolledStudentIds = const [],
    this.channels = const [],
  });

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'] ?? json['_id'] ?? '',
      code: json['code'] ?? '',
      name: json['name'] ?? '',
      instructor: json['instructor'] ?? '',
      schedule: json['schedule'] ?? '',
      location: json['location'] ?? '',
      credits: json['credits'] ?? 3,
      description: json['description'] ?? '',
      enrolledStudentIds: List<String>.from(json['enrolledStudentIds'] ?? []),
      channels: (json['channels'] as List<dynamic>?)
              ?.map((c) => Channel.fromJson(c as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'code': code,
    'name': name,
    'instructor': instructor,
    'schedule': schedule,
    'location': location,
    'credits': credits,
    'description': description,
    'enrolledStudentIds': enrolledStudentIds,
    'channels': channels.map((c) => c.toJson()).toList(),
  };
}

class Channel {
  final String id;
  final String name;
  final String icon;
  final String? createdBy;

  const Channel({
    required this.id,
    required this.name,
    this.icon = '💬',
    this.createdBy,
  });

  factory Channel.fromJson(Map<String, dynamic> json) {
    return Channel(
      id: json['id'] ?? json['_id'] ?? '',
      name: json['name'] ?? '',
      icon: json['icon'] ?? '💬',
      createdBy: json['createdBy'],
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'icon': icon,
    'createdBy': createdBy,
  };
}

File: backend/models/Course.js
const mongoose = require('mongoose');

const channelSchema = new mongoose.Schema({
  name: { type: String, required: true },
  icon: { type: String, default: '💬' },
  createdBy: { type: String, default: null }, // null = default/main channel, otherwise user name
});

const courseSchema = new mongoose.Schema({
  code:        { type: String, required: true },
  name:        { type: String, required: true },
  instructor:  { type: String, required: true },
  schedule:    { type: String, required: true },
  location:    { type: String, required: true },
  credits:     { type: Number, default: 3 },
  description: { type: String, default: '' },
  enrolledStudentIds: [{ type: String }],
  channels: [channelSchema],
}, { timestamps: true });

courseSchema.methods.toJSON = function() {
  const obj = this.toObject();
  obj.id = obj._id.toString();
  if (obj.channels) {
    obj.channels = obj.channels.map(ch => ({
      id: ch._id.toString(),
      name: ch.name,
      icon: ch.icon,
      createdBy: ch.createdBy || null,
    }));
  }
  delete obj._id;
  delete obj.__v;
  return obj;
};

module.exports = mongoose.model('Course', courseSchema);
